### PR TITLE
Add topic to collection

### PR DESCRIPTION
**Pull request by Mentessi**

### CODE REVIEW

File: app/controllers/collections_topics_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class CollectionsTopicsController < ApplicationController
+
+  def create
+    @topic = Topic.find(params[:topic_id])
+    @collection = Collection.find(params[:add_topic_to_collection])
+
+    if @topic.update(collections_topics_attributes: [{collection_id: @collection.id, topic_id: @topic.id}])
```
Comment (Neilos) [inline app/controllers/collections_topics_controller.rb]: Well done for getting this to work, but this isn't correct. You are only allowing users to assign a topic to a single collection. If that's what you wanted, then we went way overboard on the many to many association.
I'm pretty sure the associations are are correct for the desired user functionality, and it is the controller and form functionality that is too simplistic.

File: app/models/topic.rb [hunk 2]
```diff
@@ -4,4 +4,5 @@ class Topic < ApplicationRecord
   belongs_to :user
   has_many :collections_topics
   has_many :collections, through: :collections_topics
+  accepts_nested_attributes_for :collections_topics
```
Comment (Neilos) [inline app/models/topic.rb]: You're not using this. Remove.

File: app/views/topics/show.html.erb [hunk 3]
```diff
@@ -8,6 +8,19 @@
     <%= link_to 'Edit Topic', edit_topic_path(@topic.id) %>
   <% end %>
 
+  </br>
+
+  <% if user_signed_in? %>
+    <h3> Add to collection </h3>
+
+    <%= form_with model: @collections_topic, url: {controller: "collections_topics", action: "create"} do |f| %>
+     <%= select_tag 'add_topic_to_collection', options_from_collection_for_select(@collection_list, 'id', 'name') %>
```
Comment (Neilos) [inline app/views/topics/show.html.erb]: This should either be a multi-select, or checkboxes. The associations allow a topic to be assigned to multiple collections.

File: app/controllers/collections_topics_controller.rb (hunk 1 above)
Comment (Mentessi) [reply app/controllers/collections_topics_controller.rb]: I was going to add the ability to add the same topic to multiple collections later... you _can_ add topics to multiple collections this way, just manually one at a time... I'll make the changes now though, I've been looking at checkbox forms for editing collections so I think 🤞 I should be able to do it (well, get it to 'werk' anyway)

File: app/models/topic.rb (hunk 2 above)
Comment (Mentessi) [reply app/models/topic.rb]: Hmmm, dont I? I thought this was how I was adding topics to collections 😕

File: app/views/topics/show.html.erb (hunk 3 above)
Comment (Mentessi) [reply app/views/topics/show.html.erb]: Yup! Will do :)

File: app/models/topic.rb (hunk 2 above)
Comment (Neilos) [reply app/models/topic.rb]: Oh maybe you do then. 🤔 Not sure. Try taking it out and see if it breaks. You might be right

File: app/controllers/collections_topics_controller.rb (hunk 1 above)
Comment (Neilos) [reply app/controllers/collections_topics_controller.rb]: Ah, so without the `_delete` attributes it just creates new `collections_topics`, so you can do it multiple times to add multiple topics (?)

Even if that's the case, there's no way to see which collections a topic belongs to. The advantage of checkboxes or multi-selects is that it shows which collections have already been selected. At present you could end up accidentally trying to add the topic to a collection that it already belongs to, and, with the database indexes as they are, that would blow up with an exception, I think.

Comment (Mentessi) [reply app/controllers/collections_topics_controller.rb]: Yeah exactly, I just need to do it multiple times... you know it works as when you add one you are redirected to the collection show page where the topic you have just added will be listed below the collection name... that said, I do think your way with checkboxes is a better user experience... am in the process of changing the form to a checkbox one 👍

**Comment by Mentessi** [comment]
few things:

1.  collections controller handles failure via an error message so the redirect on failure is the same  as success - not sure if this is the right way of doing it?

2. I added and then removed a - unique true - for topic uniqueness in a collection as this broke the form. I couldn't get it working, technically it's not possible (from a users perspective anyway). Although I could get it working and write request specs for it, as it stopped the form working I didn't do it.

3. I haven't added any additional request specs as the ones for topics cover this already (right?)

**Comment by Neilos** [comment]
Replies to your comments:
1. It'a s question for the product owner. 😉 Is that the way they want it to work?
2. Strange. Let's merge this PR first then, and look at enforcing that at the DB separately.
3. You can alway write request specs if you want, but my policy is to only write request specs when there are circumstances not covered by the feature/system spec (e.g. security concerns) or when there are lots of different scenarios that are easier/quicker to test in a request spec. For this feature I wouldn't write a feature spec.

File: app/views/collections/show.html.erb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 <h1><%= @collection.name %></h1>
 <% @collection.topics.each do |topic| %>
   <%= topic.name %>
+  </br>
```
Comment (Neilos) [inline app/views/collections/show.html.erb]: _"Do you even CSS br?"_

**Pull request merged**